packages/types/src/type/ExtrinsicEra.ts

```diff
@@ -2,33 +2,224 @@
 
-import { AnyU8a } from '../types';
+import { assert, hexToU8a, isHex, isU8a, isObject, u8aToBn } from '@polkadot/util';
+import { AnyU8a, IExtrinsicEra } from '../types';
+import Enum from '../codec/Enum';
+import Tuple from '../codec/Tuple';
+import U8 from '../primitive/U8';
+import U8a from '../codec/U8a';
 
-import { u8aToU8a } from '@polkadot/util';
+type MortalEraValue = [U8, U8];
 
-import U8a from '../codec/U8a';
+interface MortalMethod {
+  current: number;
+  period: number;
+}
+
+const VALID_IMMORTAL = new U8a([0]);
 
 /**
  * @name ExtrinsicEra
  * @description
  * The era for an extrinsic, indicating either a mortal or immortal extrinsic
  */
-export default class ExtrinsicEra extends U8a {
+export default class ExtrinsicEra extends Enum implements IExtrinsicEra {
+  constructor (value?: any) {
+    super({
+      ImmortalEra,
+      MortalEra
+    }, ExtrinsicEra.decodeExtrinsicEra(value));
+  }
+
+  private static decodeExtrinsicEra (value: MortalMethod | Uint8Array | string = new Uint8Array()): Uint8Array | Object | undefined {
+    if (isHex(value)) {
+      return ExtrinsicEra.decodeExtrinsicEra(hexToU8a(value));
+    } else if (isU8a(value)) {
+      if (!value.length || value[0] === 0) {
+        return new Uint8Array([0, 0]);
+      } else {
+        return new Uint8Array([1, value[0], value[1]]);
+      }
+    } else if (isObject(value)) {
+      return { MortalEra: value };
+    }
+
+    throw new Error('Invalid data passed to Era');
+  }
+
+  /**
+   * @description Overide the encoded length method
+   */
+  get encodedLength (): number {
+    if (this.index === 0) {
+      return this.asImmortalEra.encodedLength;
+    } else {
+      return this.asMortalEra.encodedLength;
+    }
+  }
+
+  /**
+   * @description Returns the item as a [[ImmortalEra]]
+   */
+  get asImmortalEra (): ImmortalEra {
+    assert(this.isImmortalEra, `Cannot convert '${this.type}' via asImmortalEra`);
+
+    return this.value as ImmortalEra;
+  }
+
+  /**
+   * @description Returns the item as a [[MortalEra]]
+   */
+  get asMortalEra (): MortalEra {
+    assert(this.isMortalEra, `Cannot convert '${this.type}' via asMortalEra`);
+
+    return this.value as MortalEra;
+  }
+
+  /**
+   * @description `true` if Immortal
+   */
+  get isImmortalEra (): boolean {
+    return this.index === 0;
+  }
+
+  /**
+   * @description `true` if Mortal
+   */
+  get isMortalEra (): boolean {
+    return this.index > 0;
+  }
+
+  /**
+   * @description Encodes the value as a Uint8Array as per the parity-codec specifications
+   * @param isBare true when the value has none of the type-specific prefixes (internal)
+   */
+  toU8a (isBare?: boolean): Uint8Array {
+    if (this.index === 0) {
+      return this.asImmortalEra.toU8a(isBare);
+    } else {
+      return this.asMortalEra.toU8a(isBare);
+    }
+  }
+}
+
+/**
+ * @name ImmortalEra
+ * @description
+ * The ImmortalEra for an extrinsic
+ */
+export class ImmortalEra extends U8a {
```

Comment:
Should extend Null: https://github.com/paritytech/substrate/blob/0e6a407a13fbc5a5ad200645aed72bb8a8e528d7/core/sr-primitives/src/generic/era.rs#L32

Reply:
ImmortalEra has encodedLength of 1, while Null is 0

Reply:
So there are 2 ways to encode/decode this:

1. `Enum<ImmortalEra | MortalEra>`. Normally Enum encoding starts with 0 or 1, which denotes which enumIndex. So if it's 0, then we have ImmortalEra, and ImmortalEra should be of encodedLength=0; If it's not 0, then we don't care about the 0/1, and decode MortalEra.

2. `Enum<ImmortalEra | MortalEra>`, where we manually remove the starting 0/1 enumIndex, in this case we set encodedLength(ImmortalEra)=1 and and `ImmortalEra.toU8a() = [0]`;

Both work, but imo 1 makes a little bit more sense: we don't need the `assert(this.eq(VALID_IMMORTAL))` anymore.

Reply:
Actually, come to think of it, my proposal 1/ is not necessarily cleaner, we can keep like this. Maybe U8aFixed makes more sense though?

Reply:
For ImmortalEra, it can go both ways. but for MortalEra, the first byte is not fixed to [1], which makes it logically conflict with your proposal 1.

I think enum still make sense in this case, and in rust side it's an enum.

Reply:
Sounds good 👍